gstools/field/base.py

```diff
@@ -26,33 +27,264 @@
 )
 from gstools.tools.geometric import pos2xyz, xyz2pos
 
-__all__ = ["Field"]
+__all__ = ["Field", "Mesh"]
 
 
-class Field:
-    """A field base class for random and kriging fields ect.
+def value_type(mesh_type, shape):
+    """Determine the value type ("scalar" or "vector")"""
+    r = "scalar"
+    if mesh_type == "unstructured":
+        # TODO is this the right place for doing these checks?!
+        if len(shape) == 2 and 2 <= shape[0] <= 3:
+            r = "vector"
+    else:
+        # for very small (2-3 points) meshes, this could break
+        # a 100% solution would require dim. information.
+        if len(shape) == shape[0] + 1:
+            r = "vector"
+    return r
+
+
+class Mesh:
+    """A base class encapsulating field data.
+
+    It holds a position array, which define the spatial locations of the
+    field values.
+    It can hold multiple fields in the :any:`self.point_data` dict. This assumes
+    that each field is defined on the same positions.
+    The mesh type must also be specified.
+
+    Parameters
+    ----------
+    pos : :class:`numpy.ndarray`, optional
+        positions of the field values
+    name : :any:`str`, optional
+        key of the field values
+    values : :any:`list`, optional
+        a list of the values of the fields
+    mesh_type : :any:`str`, optional
+        the type of mesh on which the field is defined on, can be
+        * unstructured
+        * structured
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from gstools import Mesh
+    >>> pos = np.linspace(0., 100., 40)
+    >>> z = np.random.random(40)
+    >>> z2 = np.random.random(40)
+    >>> mesh = Mesh((pos,))
+    >>> mesh.add_field(z, "test_field1")
+    >>> mesh.add_field(z2, "test_field2")
+    >>> mesh.set_default_field("test_field2")
+    >>> print(mesh.field)
+
+    """
+
+    def __init__(
+        self, pos=None, name="field", values=None, mesh_type="unstructured",
+    ):
+        # mesh_type needs a special setter, therefore, `set_field_data` is not
+        # used here
+        self.mesh_type = mesh_type
+
+        # the pos/ points of the mesh
+        self._pos = pos
+
+        # data stored at each pos/ point, the "fields"
+        if values is not None:
+            self.point_data = {name: values}
+        else:
+            self.point_data = {}
+
+        # data valid for the global field
+        self.field_data = {}
+
+        self.set_field_data("default_field", name)
+
+        self.field_data["mesh_type"] = mesh_type
+
+    def set_field_data(self, name, value):
+        """Add an attribute to this instance and add it the `field_data`
+
+        This helper method is used to create attributes for easy access
+        while using this class, but it also adds an entry to the dictionary
+        `field_data`, which is used for exporting the data.
+        """
+        setattr(self, name, value)
+        self.field_data[name] = value
```

Comment:
IMO, you should avoid using `setattr` like this. In my experience, it can provide a lot of confusion for new user trying to figure out what are properties of the mesh versus data they actually used. It also can bring a lot of inconsistencies to workflows.

Also what about when a user sets a field with a space or newline character in the name? `setattr` allows those but that will not be accessible from a `.`

Reply:
The idea behind the set_field_data method is to be able to conveniently access attributes when using a Mesh class or one of its childs, but at the same time keep track of the field data properties which should be exported to VTK-files. That's what the self.field_data dictionary is for.